File: test-examples/sse-rolls-subscribe.py
#!/usr/bin/env python3
"""
SSE Rolls Subscribe example
pip install sseclient-py requests python-dotenv

Usage: python test-examples/sse-rolls-subscribe.py <clientId>
  Or set TEST_CLIENT_ID_V13 in .env.test
"""

import sys
import os
import json

try:
    import sseclient
    import requests
    from dotenv import load_dotenv
except ImportError as e:
    print(f"Missing dependency: {e}")
    print("Install with: pip install sseclient-py requests python-dotenv")
    sys.exit(1)

load_dotenv('.env.test')

base_url = os.environ.get('TEST_BASE_URL', 'http://localhost:3010')
api_key = os.environ.get('TEST_API_KEY', '')
client_id = sys.argv[1] if len(sys.argv) > 1 else os.environ.get('TEST_CLIENT_ID_V13') or os.environ.get('TEST_CLIENT_ID_V12')

if not client_id:
    print('Usage: python test-examples/sse-rolls-subscribe.py <clientId>')
    print('  Or set TEST_CLIENT_ID_V13 in .env.test')
    sys.exit(1)

url = f'{base_url}/rolls/subscribe'
params = {'clientId': client_id}
headers = {
    'x-api-key': api_key,
    'Accept': 'text/event-stream'
}

print(f'Connecting to {url}?clientId={client_id}...')

response = requests.get(url, params=params, headers=headers, stream=True)

if response.status_code != 200:
    print(f'Error: HTTP {response.status_code}')
    print(response.text)
    sys.exit(1)

client = sseclient.SSEClient(response)

try:
    for event in client.events():
        data = json.loads(event.data)

        if event.event == 'connected':
            print(f'Connected: {data["clientId"]}')
            print('Listening for roll events... (Ctrl+C to stop)\n')
        elif event.event == 'roll':
            user = (data.get('user') or {}).get('name', '?')
            crit = ' CRITICAL!' if data.get('isCritical') else ''
            fumble = ' FUMBLE!' if data.get('isFumble') else ''
            print(f'[{user}] {data["formula"]} = {data["rollTotal"]}{crit}{fumble}')
            if data.get('flavor'):
                print(f'  Flavor: {data["flavor"]}')
            for d in data.get('dice', []):
                results = ', '.join(
                    f'{r["result"]}{"" if r.get("active", True) else "(dropped)"}'
                    for r in d.get('results', [])
                )
                print(f'  Dice: {results} (d{d["faces"]})')
        else:
            print(f'Unknown event: {event.event} -> {data}')
except KeyboardInterrupt:
    print('\nDisconnecting...')

File: test-examples/sse-rolls-subscribe.js
// Node.js JavaScript SSE example for roll events
// Run with: node test-examples/sse-rolls-subscribe.js <clientId>

const { EventSource } = require('eventsource');
require('dotenv').config({ path: '.env.test' });

const baseUrl = process.env.TEST_BASE_URL || 'http://localhost:3010';
const apiKey = process.env.TEST_API_KEY;
const clientId = process.argv[2] || process.env.TEST_CLIENT_ID_V13 || process.env.TEST_CLIENT_ID_V12;

if (!clientId) {
  console.error('Usage: node test-examples/sse-rolls-subscribe.js <clientId>');
  console.error('  Or set TEST_CLIENT_ID_V13 in .env.test');
  process.exit(1);
}

const url = `${baseUrl}/rolls/subscribe?clientId=${clientId}`;
console.log(`Connecting to ${url}...`);

// eventsource v4 uses a custom fetch function to inject headers
const eventSource = new EventSource(url, {
  fetch: (input, init) => fetch(input, {
    ...init,
    headers: { ...init?.headers, 'x-api-key': apiKey }
  })
});

eventSource.addEventListener('connected', (event) => {
  const data = JSON.parse(event.data);
  console.log('Connected:', data.clientId);
  console.log('Listening for roll events... (Ctrl+C to stop)\n');
});

eventSource.addEventListener('roll', (event) => {
  const roll = JSON.parse(event.data);
  const dice = roll.dice?.map(d =>
    `${d.results.map(r => `${r.result}${r.active ? '' : '(dropped)'}`).join(', ')} (d${d.faces})`
  ).join(' + ') || '';
  console.log(`[${roll.user?.name}] ${roll.formula} = ${roll.rollTotal}${roll.isCritical ? ' CRITICAL!' : ''}${roll.isFumble ? ' FUMBLE!' : ''}`);
  if (roll.flavor) console.log(`  Flavor: ${roll.flavor}`);
  if (dice) console.log(`  Dice: ${dice}`);
});

eventSource.onerror = (error) => {
  console.error('SSE error:', error);
  eventSource.close();
  process.exit(1);
};

process.on('SIGINT', () => {
  console.log('\nDisconnecting...');
  eventSource.close();
  process.exit(0);
});
